# Ragnarok Mobile Review Scraping

In [5]:
# Will be using this to get information on each app in the Google Play Store based on appid
# Information can be found here https://github.com/JoMingyu/google-play-scraper

!pip install 'google-play-scraper'

In [1]:
# Will be using this to get information on each app in the Apple App Store based on appid
# Information can be found here https://github.com/cowboy-bebug/app-store-scraper

pip install app-store-scraper

     |████████████████████████████████| 58 kB 5.3 MB/s eta 0:00:011
  Attempting uninstall: requests
    Found existing installation: requests 2.22.0
    Uninstalling requests-2.22.0:
      Successfully uninstalled requests-2.22.0
Note: you may need to restart the kernel to use updated packages.


In [2]:
import requests
import datetime
from datetime import datetime
import time
import pandas as pd

==================================================================================================================

# Getting general information about the game from the Google Play Store

- Global ROM = "https://play.google.com/store/apps/details?id=com.gravity.romNAg"
- SEA ROM = "https://play.google.com/store/apps/details?id=com.gravity.romg"
- EU ROM = "https://play.google.com/store/apps/details?id=com.gravity.romEUg"

In [3]:
# This will do most of the work for us.

from google_play_scraper import app

# Doing a test run.
result = app(
    'com.gravity.romNAg',
    lang='en',
    country='us'
)

In [4]:
# Checking what I get in the dictionary.

result

{'title': 'Ragnarok M: Eternal Love(ROM)',
 'description': 'Dear Adventurers,\r\n\r\n"Ragnarok M: Eternal Love" has been with you for more than a thousand days. Together we\'ve experienced and witnessed many stories about gathering and parting. Now, a new era is coming!\r\n\r\nTo welcome the coming new era, various teasers about this new release can be seen. A brand new map "Moon Lake" and new Main Plot Stories will lead you to a secret about the moon and rainbow. The new multi-player gameplay mode will brings you a more exciting team confrontation. And the new Free Angle mode will improve your overall adventure experience!\r\n\r\nGet ready and bring friends along. Let\'s enter the new era of greater freedom and beauty!\r\n\r\nEP 7.5 [Moon\'s Echo] official release! Follow Fate\'s new guidance. Look for Traces of Rainbow Light!\r\n\r\n-New map [Moon Lake] unlocks a new chapter of destiny\r\nIn a faraway galaxy, the moon lingered alone outside Yggdrasill. New map [Moon Lake] and server 

In [5]:
# Checking which keys I want to keep to use as features.

result.keys()

dict_keys(['title', 'description', 'descriptionHTML', 'summary', 'summaryHTML', 'installs', 'minInstalls', 'score', 'ratings', 'reviews', 'histogram', 'price', 'free', 'currency', 'sale', 'saleTime', 'originalPrice', 'saleText', 'offersIAP', 'inAppProductPrice', 'size', 'androidVersion', 'androidVersionText', 'developer', 'developerId', 'developerEmail', 'developerWebsite', 'developerAddress', 'privacyPolicy', 'developerInternalID', 'genre', 'genreId', 'icon', 'headerImage', 'screenshots', 'video', 'videoImage', 'contentRating', 'contentRatingDescription', 'adSupported', 'containsAds', 'released', 'updated', 'version', 'recentChanges', 'recentChangesHTML', 'comments', 'appId', 'url'])

In [6]:
featurekeys = ['title', 'installs', 'minInstalls', 'score', 'ratings', 'reviews', 'histogram', 'price', 'free', 'offersIAP', 'size', 'genre', 'genreId', 'released', 'contentRating', 'appId']

In [7]:
# Filtering dictionary so I only get results I want.

filteredresult = dict((k, result[k]) for k in featurekeys if k in result)
filteredresult

{'title': 'Ragnarok M: Eternal Love(ROM)',
 'installs': '1,000,000+',
 'minInstalls': 1000000,
 'score': 3.8420374,
 'ratings': 77426,
 'reviews': 38075,
 'histogram': [15575, 3324, 4942, 7498, 46087],
 'price': 0,
 'free': True,
 'offersIAP': True,
 'size': '71M',
 'genre': 'Role Playing',
 'genreId': 'GAME_ROLE_PLAYING',
 'released': 'Jan 8, 2019',
 'contentRating': 'Teen',
 'appId': 'com.gravity.romNAg'}

In [8]:
# Now to make a function that gives me the filtered results.

def givemeinfo(appid):
    result = app(
    appid,
    lang='en',
    country='us')
    
    featurekeys = ['title', 'installs', 'minInstalls', 'score', 'ratings', 'reviews', 'histogram', 'price', 'free', 'offersIAP', 'size', 'genre', 'genreId', 'released', 'contentRating', 'appId']
    
    filteredresult = dict((k, result[k]) for k in featurekeys if k in result)
    
    return filteredresult

In [9]:
print(givemeinfo("com.gravity.romNAg"))
print('--------------------------------')
print(givemeinfo("com.gravity.romg"))
print('--------------------------------')
print(givemeinfo("com.gravity.romEUg"))

{'title': 'Ragnarok M: Eternal Love(ROM)', 'installs': '1,000,000+', 'minInstalls': 1000000, 'score': 3.8420374, 'ratings': 77426, 'reviews': 38075, 'histogram': [15575, 3324, 4942, 7498, 46087], 'price': 0, 'free': True, 'offersIAP': True, 'size': '71M', 'genre': 'Role Playing', 'genreId': 'GAME_ROLE_PLAYING', 'released': 'Jan 8, 2019', 'contentRating': 'Teen', 'appId': 'com.gravity.romNAg'}
--------------------------------
{'title': 'Ragnarok M: Eternal Love', 'installs': '5,000,000+', 'minInstalls': 5000000, 'score': 3.93567, 'ratings': 305201, 'reviews': 135427, 'histogram': [57621, 8827, 18584, 30700, 189469], 'price': 0, 'free': True, 'offersIAP': True, 'size': '69M', 'genre': 'Role Playing', 'genreId': 'GAME_ROLE_PLAYING', 'released': None, 'contentRating': 'Teen', 'appId': 'com.gravity.romg'}
--------------------------------
{'title': 'Ragnarok M: Eternal Love EU', 'installs': '100,000+', 'minInstalls': 100000, 'score': 3.9823413, 'ratings': 15264, 'reviews': 5677, 'histogram':

## For the date [12/20/2020] at 4:35 PM EST.
#### - We see that there are 38075 Global reviews
#### - We see that there are 135427 SEA reviews
#### - We see that there are 5677 EU reviews

------------
It seems more favorable to not include EU reviews for two reasons:
- There are a lack of reviews to work with
- It is not on the same version of the game as Global and SEA. It is behind by one episode (one major update)


In [10]:
# Now to get a list of all reviews for this game.

from google_play_scraper import Sort, reviews_all

In [11]:
# filter only the results that we want.

def givemereviews(appids):
    
    # Parts of the review that we are interested in
    keys = {'content', 'score', 'thumbsUpCount', 'reviewCreatedVersion', 'replyContent', 'at'}
    
    # The function itself
    res = reviews_all(
    appids,
    sleep_milliseconds=0, 
    lang='en', 
    country='us', 
    sort=Sort.NEWEST, 
    filter_score_with=None 
)
    # Filtering the results
    filteredresults = [{k:str(v) for k,v in i.items() if k in keys} for i in res]
    for item in filteredresults:
        item.update({"appId":appids})
    return filteredresults

In [12]:
starttime = datetime.now()

reviewlistglobal = givemereviews('com.gravity.romNAg')

print(datetime.now() - starttime)
print('------------')
print(len(reviewlistglobal))

0:00:09.146287
------------
6470


In [14]:
starttime = datetime.now()

reviewlistsea = givemereviews('com.gravity.romg')

print(datetime.now() - starttime)
print('------------')
print(len(reviewlistsea))

0:02:15.728508
------------
72164


In [15]:
starttime = datetime.now()

reviewlisteu = givemereviews('com.gravity.romEUg')

print(datetime.now() - starttime)
print('------------')
print(len(reviewlisteu))

0:00:01.740360
------------
573


==================================================================================================================

# Getting general information about the game from the Apple App Store

In [20]:
from app_store_scraper import AppStore
from pprint import pprint

In [30]:
countries = ['us', 'ca', 'br', 'au']
applereviews = []

for country in countries:
    test = AppStore(country=country, app_name="ragnarok-m-eternal-love", app_id=1444739251)
    test.review(sleep=8)
    applereviews.extend(test.reviews)
    
len(applereviews)

2021-01-03 04:42:05,771 [INFO] Base - Initialised: AppStore('us', 'ragnarok-m-eternal-love', 1444739251)
2021-01-03 04:42:05,772 [INFO] Base - Ready to fetch reviews from: https://apps.apple.com/us/app/ragnarok-m-eternal-love/id1444739251
2021-01-03 04:42:13,974 [INFO] Base - [id:1444739251] Fetched 20 reviews (20 fetched in total)
2021-01-03 04:42:30,300 [INFO] Base - [id:1444739251] Fetched 60 reviews (60 fetched in total)
2021-01-03 04:42:46,649 [INFO] Base - [id:1444739251] Fetched 100 reviews (100 fetched in total)
2021-01-03 04:43:02,917 [INFO] Base - [id:1444739251] Fetched 140 reviews (140 fetched in total)
2021-01-03 04:43:19,166 [INFO] Base - [id:1444739251] Fetched 180 reviews (180 fetched in total)
2021-01-03 04:43:35,421 [INFO] Base - [id:1444739251] Fetched 220 reviews (220 fetched in total)
2021-01-03 04:43:51,676 [INFO] Base - [id:1444739251] Fetched 260 reviews (260 fetched in total)
2021-01-03 04:44:07,931 [INFO] Base - [id:1444739251] Fetched 300 reviews (300 fetched

2568

==================================================================================================================

## Compressing files for use in other notebooks
We will only compress the global and sea files.

In [16]:
# We will use bz2 to compress pickle files to 1/4th the size.

import pickle
import bz2

# This is to compress it.
def compressed_pickle(title, data):
    with bz2.BZ2File(title + '.pbz2', 'w') as f:
        pickle.dump(data, f)
        
# This is to bring it back.        
def decompress_pickle(file):
    data = bz2.BZ2File(file, 'rb')
    data = pickle.load(data)
    return data

In [17]:
compressed_pickle('compressedlistreview', reviewlistglobal)
compressed_pickle('compressedlistreview2', reviewlistsea)

==================================================================================================================